FreeRTOS/FreeRTOS-Kernel
# Add port-optimised task selection for ARMv8-M

Add port-optimised task selection for ARMv8-M

Description
-----------
Add port-optimised task selection for ARMv8-M, patterned after the existing implementation on ARMv7-M.  Include a compile-time error for baseline implementations of ARMv8-M like CM23 (without the "main" extension described in the Architecture Reference Manual) attempting to use configPORT_OPTIMISED_TASK_SELECTION.  Those CPUs do not support the `clz` instruction used in the implementation of port-optimised task selection.

During the migration of this feature from ARMv7-M to ARMv8-M, the helper function `ucPortCountLeadingZeros()` became `ulPortCountLeadingZeros()` (returning a uint32_t, not a uint8_t) to stop forcing the compiler to generate an unnecessary `uxtb` instruction to clear out the upper 24 bits of the register.  The `clz` instruction always writes a full 32-bit value to the destination register.

Test Steps
-----------
Copy the port files from this PR into a local clone of the full FreeRTOS distribution.  Then:
- Build CM33 demo app (NXP LPC55S69) in IAR with and without configPORT_OPTIMISED_TASK_SELECTION.  Verify build is successful.  Review generated file task.s for use/non-use of `clz` in vTaskSwitchContext().
- Build CM33 demo app (NXP LPC55S69) in GCC with and without configPORT_OPTIMISED_TASK_SELECTION.  Verify build is successful.
- Build CM23 demo app (Nuvoton M2351) with configPORT_OPTIMISED_TASK_SELECTION set to 1 and verify error is generated during compilation.

Copy the port and kernel files from this PR into a local clone of https://github.com/jefftenney/LPTIM-Tick-U5/ (for STM32U585, CM33).  Then:
- Build and run with configPORT_OPTIMISED_TASK_SELECTION set to 1.  Use debugger to verify use of `clz` in disassembly for vTaskSwitchContext().
- Verify correct operation of application.

Checklist:
----------


[X] I have tested my changes. No regression in existing tests.
~[X] I have modified and/or added unit-tests to cover the code changes in this Pull Request.~ (N/A)

Related Issues
-----------
- https://forums.freertos.org/t/port-optimized-task-selection-on-cortex-m33/12521
- Closes #319

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
